PennMobile/New Group/PAC Code/PacCodeViewController.swift

```diff
@@ -0,0 +1,150 @@
+        pacCodeInfoLabel.centerXAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor).isActive = true
+    }
+    
+    func refreshPacCode() {
+        if let pacCode = pacCode {
+            var counter = 0
+            for pacCodeDigit in pacCode {
+                quadDigitLabel[counter].text = "\(pacCodeDigit)"
+                counter += 1
+            }
+        } else {
+            for box in quadDigitLabel {
+                box.text = ""
+            }
+        }
+    }
+}
+
+// MARK: - Local Authentication
+extension PacCodeViewController : LocalAuthentication {
+        
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        handleAuthentication(cancelText: "Go Back", reasonText: "Authenticate to see your PAC Code")
+    }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        pacCode = nil
+        refreshPacCode()
+    }
+    
+    func handleAuthenticationSuccess() {
+        if let pacCode = getPacCode() {
+            self.pacCode = pacCode
+        } else {
+            self.showAlert(withMsg: "Please login to use this feature", title: "Login Error", completion: { self.navigationController?.popViewController(animated: true)} )
+        }
+        
+        refreshPacCode()
+    }
+    
+    func handleAuthenticationFailure() {
+        self.navigationController?.popViewController(animated: true)
+    }
+        
+}
+
+// MARK: - PAC Code Refreshing
+extension PacCodeViewController {
+    @objc func refreshButtonTapped() {
+        let message = "Has there been a change to your PAC Code? Would you like Penn Mobile to update your information?"
+        let alert = UIAlertController(title: "Update PAC Code", message: message, preferredStyle: .alert)
+        
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
+        
+        let refreshPacCode = UIAlertAction(title: "Yes", style: .default, handler: { (UIAlertAction) in
+            self.showActivity()
+            PacCodeNetworkManager.instance.getPacCode(callback: self.handleNetworkPacCodeRefreshCompletion(_:))
+        })
+        
+        alert.addAction(refreshPacCode)
+        alert.addAction(cancelAction)
+        
+        present(alert, animated: true, completion: nil)
+    }
+    
+    fileprivate func handleNetworkPacCodeRefreshCompletion(_ pacCode: String?) {
+        DispatchQueue.main.async {
+            self.hideActivity()
+            if let pacCode = pacCode {
+                self.savePacCode(pacCode)
+                self.pacCode = pacCode
+                self.showAlert(withMsg: "Your PAC Code has been updated.", title: "Success!", completion: self.refreshPacCode)
+            } else {
+                self.showAlert(withMsg: "Unable to access your courses. Please try again later.", title: "Uh oh!", completion: nil)
```

Comment:
Huh? I think you meant to change this.

Reply:
Sorry! I should have caught this mistake.